### PR TITLE
Migration of test_volume_create.py

### PR DESCRIPTION
**Pull request by nishith-vihar**
Migration of [test_volume_create.py](https://github.com/gluster/glusto-tests/blob/master/tests/functional/glusterd/test_volume_create.py) from glusto to redant

Updates: #292
Fixes: #562 
Signed-off-by: Nishith Vihar Sakinala nsakinal@redhat.com

### CODE REVIEW

**Comment by nishith-vihar** [comment]
Required 4 servers in the test case logic

**Comment by nik-redhat** [comment]
Rebase the PR, and add a check at the top for the servers and let's take this PR to completion.

File: common/ops/gluster_ops/volume_ops.py [hunk 1]
```diff
@@ -69,7 +69,8 @@ def setup_volume(self, volname: str, node: str, conf_hash: dict,
 
     def volume_create(self, volname: str, node: str, conf_hash: dict,
```
Comment (nik-redhat) [inline common/ops/gluster_ops/volume_ops.py:70]: A new function for this is added, `volume_create_with_custome_brick` or something of that sort, use that.

Comment (nishith-vihar) [reply common/ops/gluster_ops/volume_ops.py:70]: Ack

File: tests/functional/glusterd/test_volume_create.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        redant.logger.info("Get all the bricks of the volume")
+        bricks_list = redant.es.get_all_bricks_list(self.volume_name1)
+        if len(bricks_list) == 0:
+            raise Exception("Failed to get the brick list")
+
+        ret = redant.bring_bricks_offline(self.volume_name1, bricks_list[0:2])
+        if not ret:
+            raise Exception("Failed to bring down the bricks")
+
+        ret = redant.volume_start(self.volume_name1, self.server_list[0],
```
Comment (nik-redhat) [inline tests/functional/glusterd/test_volume_create.py:65]: ret not needed

File: tests/functional/glusterd/test_volume_create.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        # create volume with previously used bricks and different volume name
+        self.volume_type2 = 'dist'
+        self.volume_name2 = f"{self.test_name}-{self.volume_type2}-2"
+        conf_dict = self.vol_type_inf[self.conv_dict[self.volume_type2]]
+        try:
+            redant.volume_create_with_custom_bricks(self.volume_name2,
+                                                    self.server_list[0],
+                                                    conf_dict, brick_cmd,
+                                                    brick_dict)
+            raise Exception("Unexpected: Successfully created the volume "
```
Comment (nik-redhat) [inline tests/functional/glusterd/test_volume_create.py:83]: You are raising an exception here and handling it the next line by saying `expected` which is not true.
Instead, use the excep flag and check the error_code in the return value and raise an exception accordingly.

File: tests/functional/glusterd/test_volume_create.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+        except Exception:
+            redant.logger.info("Expected: Failed to create a volume "
+                               "with previously used bricks.")
+
+        # create a volume with already existing volume name
+        try:
+            redant.setup_volume(self.volume_name1, self.server_list[0],
+                                conf_dict, self.server_list,
+                                self.brick_roots, True)
+            raise Exception("Unexpected: Successfully created the volume with "
```
Comment (nik-redhat) [inline tests/functional/glusterd/test_volume_create.py:94]: Same as above

File: tests/functional/glusterd/test_volume_create.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+  Test glusterd behavior with the gluster volume create command
+"""
+
+import random
+from tests.d_parent_test import DParentTest
+
+
+# disruptive;dist
+class TestCase(DParentTest):
+
```
Comment (nik-redhat) [inline tests/functional/glusterd/test_volume_create.py:28]: Use the setup_test function to override the volume creation in the parent_run_test.

Comment (nik-redhat) [reply tests/functional/glusterd/test_volume_create.py:28]: Also, we can simply make it `disruptive`, we don't need to create a dist volume in the parent_run_test

File: tests/functional/glusterd/test_volume_create.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                            "with previously used bricks")
+        except Exception:
+            redant.logger.info("Expected: Failed to create a volume "
+                               "with previously used bricks.")
+
+        # create a volume with already existing volume name
+        try:
+            redant.setup_volume(self.volume_name1, self.server_list[0],
+                                conf_dict, self.server_list,
+                                self.brick_roots, True)
```
Comment (nik-redhat) [inline tests/functional/glusterd/test_volume_create.py:93]: No need for `force`, also, pass `False` to except to get the return value accordingly.

File: tests/functional/glusterd/test_volume_create.py [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+        try:
+            redant.volume_create_with_custom_bricks(self.volume_name3,
+                                                    self.server_list[0],
+                                                    conf_dict, brick_cmd,
+                                                    brick_dict)
+            raise Exception("Unexpected: Successfully created the volume "
+                            "with non existing brick path")
+        except Exception:
+            redant.logger.info("Expected: Failed to create the volume with "
+                               "non existing brick path")
```
Comment (nik-redhat) [inline tests/functional/glusterd/test_volume_create.py:120]: Same as above

File: tests/functional/glusterd/test_volume_create.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+        # cluster should fail
+        try:
+            redant.setup_volume(self.volume_name3, self.server_list[0],
+                                conf_dict, self.server_list,
+                                self.brick_roots, True)
+            raise Exception("Unexpected: Successfully created the volume "
+                            "with bricks which are part of another cluster")
+        except Exception:
+            redant.logger.info("Expected: Failed to create the volume with "
+                               "bricks which are part of another cluster")
```
Comment (nik-redhat) [inline tests/functional/glusterd/test_volume_create.py:152]: Same as above.
No need for `force` and use except flag for return value

File: tests/functional/glusterd/test_volume_create.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+        conf_dict = self.vol_type_inf[self.conv_dict[self.volume_type4]]
+        try:
+            redant.setup_volume(self.volume_name4, self.server_list[0],
+                                conf_dict, self.server_list,
+                                self.brick_roots, True)
+            raise Exception("Unexpected: Successfully created the volume "
+                            "with brick whose node is down")
+        except Exception:
+            redant.logger.info("Expected: Failed to create the volume with "
+                               "brick whose node is down")
```
Comment (nik-redhat) [inline tests/functional/glusterd/test_volume_create.py:178]: Same as above

**Review by nik-redhat — CHANGES_REQUESTED** [review]

File: tests/functional/glusterd/test_volume_create.py (hunk 2 above)
Comment (nishith-vihar) [reply tests/functional/glusterd/test_volume_create.py:65]: Ack

File: tests/functional/glusterd/test_volume_create.py (hunk 3 above)
Comment (nishith-vihar) [reply tests/functional/glusterd/test_volume_create.py:83]: Ack

File: tests/functional/glusterd/test_volume_create.py (hunk 4 above)
Comment (nishith-vihar) [reply tests/functional/glusterd/test_volume_create.py:94]: Ack

File: tests/functional/glusterd/test_volume_create.py (hunk 9 above)
Comment (nishith-vihar) [reply tests/functional/glusterd/test_volume_create.py:178]: Ack

File: tests/functional/glusterd/test_volume_create.py (hunk 7 above)
Comment (nishith-vihar) [reply tests/functional/glusterd/test_volume_create.py:120]: Ack

File: tests/functional/glusterd/test_volume_create.py (hunk 6 above)
Comment (nishith-vihar) [reply tests/functional/glusterd/test_volume_create.py:93]: I don't think we need to set excep flag to False as we are not using the ret value anywhere

File: tests/functional/glusterd/test_volume_create.py (hunk 8 above)
Comment (nishith-vihar) [reply tests/functional/glusterd/test_volume_create.py:152]: I dont think we would need to set the excep flag to False here also. same as above

File: tests/functional/glusterd/test_volume_create.py (hunk 5 above)
Comment (nishith-vihar) [reply tests/functional/glusterd/test_volume_create.py:28]: I think it would be just enough to remove the dist volume just make it disruptive

Comment (nik-redhat) [reply tests/functional/glusterd/test_volume_create.py:28]: Yes, that would do as well.

File: tests/functional/glusterd/test_volume_create.py (hunk 6 above)
Comment (nik-redhat) [reply tests/functional/glusterd/test_volume_create.py]: The only reason to use `False` is that if we except the Exception here and just log it then we will have to logs now at the ERROR level done by the abstract ops and another here at INFO level saying it is expected which will create disparities.

File: tests/functional/glusterd/test_volume_create.py [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+        self.volume_type2 = 'dist'
+        self.volume_name2 = f"{self.test_name}-{self.volume_type2}-2"
+        conf_dict = self.vol_type_inf[self.conv_dict[self.volume_type2]]
+        excep = True
+        try:
+            redant.volume_create_with_custom_bricks(self.volume_name2,
+                                                    self.server_list[0],
+                                                    conf_dict, brick_cmd,
+                                                    brick_dict)
+            excep = False
```
Comment (nik-redhat) [inline tests/functional/glusterd/test_volume_create.py:84]: Hey this is not the correct way of using it, we need to pass it in the function itself, check the function definition for the usage.

**Review by nik-redhat — APPROVED** [review]

**Review by nik-redhat — APPROVED** [review]

**Pull request merged**